### PR TITLE
Update to Discord.js 12, add delete node, add edit feature

### PR DESCRIPTION
**Pull request by Markoudstaal**
**No breaking changes.**

- Refactor codebase for use with DiscordJS v12.
- Adds discordDeleteMessage for message deleting.
- discordSendMessage is now also able to edit messages.
- Updated node documentation in node-red.

### CODE REVIEW

File: discord/discordMessage.html
Comment (jorisvddonk) [inline discord/discordMessage.html:35]: spelling nitpick: recieve -> receive

**Comment by jorisvddonk** [comment]
Hi! Thanks so much for the PR! 👍 
A quick readthrough doesn't reveal anything particularly worrisome, but I'll have to give it a try myself and I'd like to give it a closer look as well. Hopefully I'll be able to get that done sometime in the next week.

**Comment by Markoudstaal** [comment]
> Hi! Thanks so much for the PR! 👍 
> 
> A quick readthrough doesn't reveal anything particularly worrisome, but I'll have to give it a try myself and I'd like to give it a closer look as well. Hopefully I'll be able to get that done sometime in the next week.

Hey Joris!
Any chance you've been able to take a look yet?

**Comment by ozdeadmeat** [comment]
Has this been integrated yet? This has a bunch of new features that I need.

**Comment by jorisvddonk** [comment]
Hi! Sorry for the delay; I haven't been able to look at it before my vacation, and now I'm no longer at home with limited connectivity and power and without my usual workstation, so taking a look is going to be be difficult until I get some good connectivity and a chance to figure out how to make things work on my Chromebook.

In the meantime I recommend using a git:// URL in your package JSON, or uploading your own module to NPM.

File: discord/discordSendMessage.html
Comment (jorisvddonk) [inline discord/discordSendMessage.html:65]: 3 `p` tags need to be closed here

File: discord/discordSendMessage.js
Comment (jorisvddonk) [inline discord/discordSendMessage.js:38]: `undefined ||` is superfluous here (the right hand side will always be used). I recommend changing this to `var msgId = msg.id;` instead.

File: package.json
Comment (jorisvddonk) [inline package.json:3]: As this is likely a breaking change for users that use the `discordClient` node due to the `discord.js` API updates (see [here](https://v12.discordjs.guide/additional-info/changes-in-v12.html) for breaking changes), according to SemVer we should rev `node-red-contrib-discord` to `5.0`, instead.

File: discord/discordDeleteMessage.html
Comment (jorisvddonk) [inline discord/discordDeleteMessage.html:48]: Node-red itself provides the `delay` node for this. Is that node sufficient? If not, why should delay functionality instead be implemented in `discordDeleteMessage`?

File: discord/discordDeleteMessage.html
Comment (jorisvddonk) [inline discord/discordDeleteMessage.html:54]: `p` tag here needs to be closed

File: discord/discordDeleteMessage.html
Comment (jorisvddonk) [inline discord/discordDeleteMessage.html:15]: The `discordSendMessage` provides a mechanism for setting the channel id within the node itself as a configuration parameter, whilst `discordDeleteMessage` does not. Either works for me, though the lack of consistency here is not ideal, and something that I think would be good to address.

The way I see it, we have two options here:
1. Remove the ability in `discordSendMessage` to specify the channel in the node config. This would be a breaking change that needs to be mentioned in the changelog.
2. Add the ability in `discordDeleteMessage` to specify the channel id, in the same way as the `discordSendMessage` node allows it.

Personally I'd lean to option 2 to avoid breaking changes, but again, I'm open to either, especially considering this changeset already brings breaking changes with it.

File: CHANGELOG.md
Comment (jorisvddonk) [inline CHANGELOG.md:3]: `4.1` -> `5.0` due to breaking changes in upstream Discord.js

File: CHANGELOG.md
Comment (jorisvddonk) [inline CHANGELOG.md:8]: This is much appreciated. Thanks! 🙏

File: discord/discordDeleteMessage.js
Comment (jorisvddonk) [inline discord/discordDeleteMessage.js:23]: This will throw a referenceError as `msgID` is not defined.
`msgID.id` -> `msgId.id`

File: discord/discordDeleteMessage.html
Comment (jorisvddonk) [reply discord/discordDeleteMessage.html:48]: Discord.js v13 also no longer supports this (see [the v12 -> v13 upgrade documentation entry for this](https://discordjs.guide/additional-info/changes-in-v13.html#message-delete)) so I recommend deleting this option.

File: discord/discordDeleteMessage.js
Comment (jorisvddonk) [inline discord/discordDeleteMessage.js:41]: Since there's a return on line 48, clearing the `channel` local variable is superfluous

**Review by jorisvddonk — CHANGES_REQUESTED** [review]
Neat stuff! I've tested it out thoroughly on my own system using node-red v1.3.6, and most of the functionality works perfectly! My apologies for the tardiness in my review; I wasn't able to take a good detailed look at any of this whilst on holiday.

I've left a few comments regarding some changes I'd like to see before I'm willing to merge this in and cut a release. I should be able to take care of these and merge things in later this week or early next week, but otherwise you're also free to update the PR yourself if you wish.

Since node-red v2 has been released in the meantime, my plans are to cut a release based on this PR for node-red v1.x, followed by a release containing any updates (if necessary) for node-red v2, so that users will be able to make use of these changes whilst still choosing to stick with node-red v1.x.

Thanks again for the updates! I really appreciate it. 👍

**Comment by jorisvddonk** [comment]
(maintainer notice: this fixes #28 for users that have Discord.js 12 installed)

**Comment by jorisvddonk** [comment]
(maintainer notice: this implements #16 )

**Comment by jorisvddonk** [comment]
(maintainer notice: this fixes #8 for Discord.js 12 users)

**Comment by Markoudstaal** [comment]
Hey Joris, since we needed to have the new functionality up on npm I created a new repository. You can find it here: https://github.com/Markoudstaal/node-red-contrib-discord-advanced
Ofcourse I do mention your work in there but we are quite a few features, nodes and pushes ahead now and there's already some people using it in node-red. Therefor I don't think merging would be the way to go anymore.

**Comment by jorisvddonk** [comment]
> Hey Joris, since we needed to have the new functionality up on npm I created a new repository. You can find it here: https://github.com/Markoudstaal/node-red-contrib-discord-advanced
> Ofcourse I do mention your work in there but we are quite a few features, nodes and pushes ahead now and there's already some people using it in node-red. Therefor I don't think merging would be the way to go anymore.

Great! That's probably the best way to go forward.
I've applied the requested changes and merged with master, and will probably cut a release later, as the update to Discord.js 12 is pretty sweet and it resolves some issues people have been having.
Thanks again for the PR, and good luck with your fork!

**Pull request merged**
